cohttp-lwt/src/server.ml

```diff
@@ -111,52 +111,60 @@ module Make (IO : S.IO) = struct
                 `Response rsp))
       (fun () -> Body.drain_body body)
 
-  let handle_response ~keep_alive oc res body handle_client =
+  let handle_response ~keep_alive oc res body conn_closed handle_client =
     IO.catch (fun () ->
         let flush = Response.flush res in
         Response.write ~flush
           (fun writer -> Body.write_body (Response.write_body writer) body)
-          res oc
-        >>= fun () -> if keep_alive then handle_client oc else Lwt.return_unit)
+          res oc)
     >>= function
-    | Ok () -> Lwt.return_unit
+    | Ok () ->
+        if keep_alive then handle_client oc
```

Comment:
If `handle_client` fails to read here we will never drain the body. This was related to an issue that led to a DOS: https://github.com/mirage/ocaml-cohttp/pull/977

Reply:
I am not against this change, but we need to make it in such a way as to not re-open https://github.com/mirage/ocaml-cohttp/issues/978

The rule of thumb was that there shoudl be no infinite body, and if tje server has it, there should be a server-side way to prevent it. Maybe in this case it could be more suitable to use https://github.com/mirage/ocaml-cohttp/blob/master/cohttp-server-lwt-unix/src/cohttp_server_lwt_unix.mli which has a finer control on the body?

Reply:
(I know, it is rather undocumented unfortunately, it is still a work in progress)

Reply:
cohttp-server-lwt-unix looks promising and clearly it may be the way, going forward. 

That being said, I think this patch does not reopen #978: If `IO.catch [...]`returns `Ok ()`, it means that `Response.write [...]` succeed which means that the `Body.write_body [...]` have all been successfully completed so this body of that response fully consumed and there is no more things to drain here, isn't it ?

Reply:
You are right, I went too fast there. I think it is fine. I wonder if we can add a test case, some kind of update of https://github.com/mirage/ocaml-cohttp/pull/977/files#diff-efd414a623c8f96d3b50c273441dd9a953d0d721a3369f4588ba01a3d1a0710d perhaps using the example you linked.